sceptre/template.py

```diff
@@ -319,6 +312,17 @@ def get_boto_call_parameter(self):
         else:
             return {"TemplateBody": self.body}
 
+    def _bucket_region(self, bucket_name):
+        return self.connection_manager.call(
+            service="s3",
+            command="get_bucket_location",
+            kwargs={"Bucket": bucket_name}
+        ).get('LocationConstraint')
```

Comment:
[aws docs](https://docs.aws.amazon.com/cli/latest/reference/s3api/get-bucket-location.html) says.. "Returns the Region the bucket resides in. **You set the bucket's Region using the LocationConstraint request parameter in a CreateBucket request**."  Therefore this request can return `None`,  it's doing that for all my buckets because I don't set `LocationConstraint` when i create buckets. 

```
aws s3api get-bucket-location --bucket bootstrap-awss3cloudformationbucket-l0axns15gl9h
None
```

This will cause sceptre to fail when setting `template_bucket_name: bootstrap-awss3cloudformationbucket-l0axns15gl9h` in config.yaml

<details>
<summary>sceptre run log</summary>

		[2021-05-01 08:45:15] - prod/noop - Launching Stack
		[2021-05-01 08:45:15] - No cloudformation client found, creating one...
		[2021-05-01 08:45:15] - Getting Boto3 session
		[2021-05-01 08:45:15] - No Boto3 session found, creating one...
		[2021-05-01 08:45:15] - Using cli credentials...
		[2021-05-01 08:45:15] - Using credential set from assume-role: {'AccessKeyId': '****************', 'SecretAccessKey': '************************************', 'Region': 'us-east-1'}
		[2021-05-01 08:45:15] - Boto3 session created
		[2021-05-01 08:45:16] - prod/noop - Stack is in the PENDING state
		[2021-05-01 08:45:16] - prod/noop - Creating Stack
		[2021-05-01 08:45:16] - noop - Uploading template to S3...
		[2021-05-01 08:45:16] - noop - Attempting to find template bucket 'bootstrap-awss3cloudformationbucket-l0axns15gl9h'
		[2021-05-01 08:45:16] - No s3 client found, creating one...
		[2021-05-01 08:45:16] - Getting Boto3 session
		[2021-05-01 08:45:16] - noop - Found template bucket 'bootstrap-awss3cloudformationbucket-l0axns15gl9h'
		[2021-05-01 08:45:17] - noop - Uploading template to: 's3://bootstrap-awss3cloudformationbucket-l0axns15gl9h/testing/prod/noop/2021-05-01-15-45-13-362578Z.json'
		Traceback (most recent call last):
			File "/usr/local/var/pyenv/versions/py38-sceptredev/bin/sceptre", line 8, in <module>
				sys.exit(cli())
			File "/usr/local/var/pyenv/versions/3.8.6/envs/py38-sceptredev/lib/python3.8/site-packages/click/core.py", line 829, in __call__
				return self.main(*args, **kwargs)
			File "/usr/local/var/pyenv/versions/3.8.6/envs/py38-sceptredev/lib/python3.8/site-packages/click/core.py", line 782, in main
				rv = self.invoke(ctx)
			File "/usr/local/var/pyenv/versions/3.8.6/envs/py38-sceptredev/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
				return _process_result(sub_ctx.command.invoke(sub_ctx))
			File "/usr/local/var/pyenv/versions/3.8.6/envs/py38-sceptredev/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
				return ctx.invoke(self.callback, **ctx.params)
			File "/usr/local/var/pyenv/versions/3.8.6/envs/py38-sceptredev/lib/python3.8/site-packages/click/core.py", line 610, in invoke
				return callback(*args, **kwargs)
			File "/usr/local/var/pyenv/versions/3.8.6/envs/py38-sceptredev/lib/python3.8/site-packages/click/decorators.py", line 21, in new_func
				return f(get_current_context(), *args, **kwargs)
			File "/usr/local/var/pyenv/versions/3.8.6/envs/py38-sceptredev/lib/python3.8/site-packages/sceptre/cli/helpers.py", line 37, in decorated
				return func(*args, **kwargs)
			File "/usr/local/var/pyenv/versions/3.8.6/envs/py38-sceptredev/lib/python3.8/site-packages/sceptre/cli/launch.py", line 38, in launch_command
				responses = plan.launch()
			File "/usr/local/var/pyenv/versions/3.8.6/envs/py38-sceptredev/lib/python3.8/site-packages/sceptre/plan/plan.py", line 130, in launch
				return self._execute(*args)
			File "/usr/local/var/pyenv/versions/3.8.6/envs/py38-sceptredev/lib/python3.8/site-packages/sceptre/plan/plan.py", line 40, in _execute
				return executor.execute(*args)
			File "/usr/local/var/pyenv/versions/3.8.6/envs/py38-sceptredev/lib/python3.8/site-packages/sceptre/plan/executor.py", line 55, in execute
				stack, status = future.result()
			File "/usr/local/var/pyenv/versions/3.8.6/lib/python3.8/concurrent/futures/_base.py", line 432, in result
				return self.__get_result()
			File "/usr/local/var/pyenv/versions/3.8.6/lib/python3.8/concurrent/futures/_base.py", line 388, in __get_result
				raise self._exception
			File "/usr/local/var/pyenv/versions/3.8.6/lib/python3.8/concurrent/futures/thread.py", line 57, in run
				result = self.fn(*self.args, **self.kwargs)
			File "/usr/local/var/pyenv/versions/3.8.6/envs/py38-sceptredev/lib/python3.8/site-packages/sceptre/plan/executor.py", line 62, in _execute
				result = getattr(actions, self.command)(*args)
			File "/usr/local/var/pyenv/versions/3.8.6/envs/py38-sceptredev/lib/python3.8/site-packages/sceptre/hooks/__init__.py", line 104, in decorated
				response = func(self, *args, **kwargs)
			File "/usr/local/var/pyenv/versions/3.8.6/envs/py38-sceptredev/lib/python3.8/site-packages/sceptre/plan/actions.py", line 201, in launch
				status = self.create()
			File "/usr/local/var/pyenv/versions/3.8.6/envs/py38-sceptredev/lib/python3.8/site-packages/sceptre/hooks/__init__.py", line 104, in decorated
				response = func(self, *args, **kwargs)
			File "/usr/local/var/pyenv/versions/3.8.6/envs/py38-sceptredev/lib/python3.8/site-packages/sceptre/plan/actions.py", line 74, in create
				self.stack.template.get_boto_call_parameter())
			File "/usr/local/var/pyenv/versions/3.8.6/envs/py38-sceptredev/lib/python3.8/site-packages/sceptre/template.py", line 310, in get_boto_call_parameter
				url = self.upload_to_s3()
			File "/usr/local/var/pyenv/versions/3.8.6/envs/py38-sceptredev/lib/python3.8/site-packages/sceptre/template.py", line 228, in upload_to_s3
				bucket_name, bucket_region, self._domain_from_region(bucket_region), bucket_key
			File "/usr/local/var/pyenv/versions/3.8.6/envs/py38-sceptredev/lib/python3.8/site-packages/sceptre/template.py", line 324, in _domain_from_region
				return "com.cn" if region.startswith("cn-") else "com"
		AttributeError: 'NoneType' object has no attribute 'startswith'

</details>

Reply:
Hi @zaro0508, good catch, you are right, the request can return None. Fortunately, that api call should still work for us. None/null is returned [when the bucket is in `us-east-1`][1], not specifically because you didn't set the constraint. Of course, since `us-east-1` is the [default][2] when creating buckets, it does lead to this problem. Looking at the logs, it appears you are using `us-east-1`, so that checks out.

I can update the pr to default to `us-east-1` to align with the (seemingly odd) boto behavior. Thanks again for finding this. 

[1]: https://docs.aws.amazon.com/cli/latest/reference/s3api/get-bucket-location.html#output
[2]: https://docs.aws.amazon.com/AmazonS3/latest/API/API_CreateBucket.html

Reply:
Sorry for the delay. I just updated the behavior to use `us-east-1` when the request returns None. Again, this is appears to be how to interpret a None return value as [per the docs][1]. 

[1]: https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Client.get_bucket_location